### PR TITLE
SDK-781/ Add README for CAA

### PR DESCRIPTION
**Pull request by Shaik-Luqmaan**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -85,6 +88,165 @@ public class TokenGenerationUtil {
...
+
+``` java
+import com.skyflow.entities.ResponseToken;
+import com.skyflow.errors.SkyflowException;
+import com.skyflow.serviceaccount.util.Token;
+import java.io.File;
+
+public class BearerTokenWithContextGeneration {
+    public static void main(String args[]) {
+
```
Comment (thefalc) [inline README.md:105]: Remove line

File: README.md [hunk 2]
```diff
@@ -85,6 +88,165 @@ public class TokenGenerationUtil {
     }
 }
 ```
+## Service Account Bearer Token with Context Generation 
+
+The service account generated with `context_id` identifier enabled can be used to generate bearer tokens with `context`.  The token generated from this service account is valid for 60 minutes and can be used to make API calls to vault services as well as management API(s) based on the permissions of the service account.
```
Comment (thefalc) [inline README.md:93]: This first sentence "The service account generated with `context_id` identifier enabled can be used to generate bearer tokens with `context`." doesn't make sense to me.

What is a context_id identifier and as someone outside of Skyflow, how would I know what that is?

File: README.md [hunk 3]
```diff
@@ -85,6 +88,165 @@ public class TokenGenerationUtil {
...
+import com.skyflow.errors.SkyflowException;
+import com.skyflow.serviceaccount.util.Token;
+import java.io.File;
+
+public class BearerTokenWithContextGeneration {
+    public static void main(String args[]) {
+
+        String bearerToken = null;
+
+        // Generate BearerToken with context by specifying credentials.json file path
```
Comment (thefalc) [inline README.md:108]: Generate a bearer token with a context value of "abc".

File: README.md [hunk 4]
```diff
@@ -85,6 +88,165 @@ public class TokenGenerationUtil {
...
+                    .setContext("abc")
+                    .build();
+
+            bearerToken = token.getBearerToken();
+            System.out.println(bearerToken);
+        } catch (Exception e) {
+            e.printStackTrace();
+        }
+
+        // Generate BearerToken with context by specifying credentials.json as string
```
Comment (thefalc) [inline README.md:122]: Generate a bearer token using a service account key string with a context value of "abc".

File: README.md [hunk 5]
```diff
@@ -85,6 +88,165 @@ public class TokenGenerationUtil {
...
+        try {
+            String fileContents = "<YOUR_CREDENTIALS_FILE_CONTENTS_AS_STRING>";
+            BearerToken token = new BearerToken.BearerTokenBuilder()
+                    .setCredentials(fileContents)
+                    .setContext("abc")
+                    .build();
+
+            bearerToken = token.getBearerToken();
+            System.out.println(bearerToken);
+
```
Comment (thefalc) [inline README.md:132]: Remove line

File: README.md [hunk 6]
```diff
@@ -85,6 +88,165 @@ public class TokenGenerationUtil {
...
+        } catch (SkyflowException e) {
+            e.printStackTrace();
+        }
+
```
Comment (thefalc) [inline README.md:136]: Remove line

File: README.md [hunk 7]
```diff
@@ -85,6 +88,165 @@ public class TokenGenerationUtil {
...
+    }
+}
+
+```
+
+Note: 
+- Either credentials file path or credentials as string can be passed to `setCredentials` method of BearerTokenBuilder class.
+- If both credentials file path and credentials string are passed to `setCredentials` method, then setter which is at the last will be taken precedence over the other.
```
Comment (thefalc) [inline README.md:144]: If both a file path and string are passed to the `setCredentials` method, then the last method used will take precedence.

File: README.md [hunk 8]
```diff
@@ -85,6 +88,165 @@ public class TokenGenerationUtil {
...
+
+## Service Account Scoped Bearer Token Generation
+
+A service account having multiple roles can be used to generate bearer tokens with restricted access by providing the appropriate `roleID`. The token generated is valid for 60 minutes and perform opertaions with permissions associated with the role.
```
Comment (thefalc) [inline README.md:148]: The token generated is valid for 60 minutes and can only perform operations with permissions associated with the given role.

File: README.md [hunk 9]
```diff
@@ -85,6 +88,165 @@ public class TokenGenerationUtil {
...
+
+[Example](https://github.com/skyflowapi/skyflow-java/blob/master/samples/src/main/java/com/example/ScopedTokenGeneration.java):
+
+```java
+import java.io.File;
+
+public class ScopedTokenGeneration {
+    public static void main(String args[]) {
+
```
Comment (thefalc) [inline README.md:157]: Remove line

File: README.md [hunk 10]
```diff
@@ -85,6 +88,165 @@ public class TokenGenerationUtil {
...
+        String scopedToken = null;
+
+        // Generate Scoped Token  by specifying credentials.json file path
```
Comment (thefalc) [inline README.md:160]: Generate a bearer token using a service account file path scoped to a specific role.

File: README.md [hunk 11]
```diff
@@ -85,6 +88,165 @@ public class TokenGenerationUtil {
...
+        try {
+            String filePath = "<YOUR_CREDENTIALS_FILE_PATH>";
+            BearerToken token = new BearerToken.BearerTokenBuilder()
```
Comment (thefalc) [inline README.md:163]: Make sure all code is run through a linter. The spacing here looks off from the standard.

File: README.md [hunk 12]
```diff
@@ -85,6 +88,165 @@ public class TokenGenerationUtil {
...
+            scopedToken = token.getBearerToken();
+            System.out.println(scopedToken);
+        } catch (Exception e) {
+            e.printStackTrace();
+        }
+    }
+}
+```
+Note: 
+- Either credentials file path or credentials as string can be passed to `setCredentials` method of BearerTokenBuilder class.
```
Comment (thefalc) [inline README.md:177]: Update based on last comment for these

File: README.md [hunk 13]
```diff
@@ -85,6 +88,165 @@ public class TokenGenerationUtil {
...
+- If both credentials file path and credentials string are passed to `setCredentials` method, then setter which is at the last will be taken precedence over the other.
+
+## Signed Data Tokens Generation
+
+Skyflow data tokens are generated when sensitive data is inserted into the vault. These data tokens can be digitally signed with the private key of the service account credentials provided, which adds an additional layer of protection. These signed tokens can then be detokenized by passing a signed data token and a bearer token generated from service account credentials having appropriate permissions and context to detokenize API calls.
```
Comment (thefalc) [inline README.md:182]: These data tokens can be digitally signed with the private key of the service account credentials, which adds an additional layer of protection. Signed tokens can be detokenized by passing the signed data token and a bearer token generated from service account credentials. The service account must have appropriate permissions and context to detokenize the signed data tokens.

File: README.md [hunk 14]
```diff
@@ -85,6 +88,165 @@ public class TokenGenerationUtil {
...
+
+``` java
+
+import com.skyflow.errors.SkyflowException;
+import java.io.File;
+import java.util.List;
+
+public class SignedTokenGeneration {
+    public static void main(String args[]) {
+
```
Comment (thefalc) [inline README.md:195]: Remove line

File: README.md [hunk 15]
```diff
@@ -85,6 +88,165 @@ public class TokenGenerationUtil {
...
+        List<SignedDataTokenResponse> signedTokenValue;
+
+        // Generate BearerToken with context by specifying credentials.json file path
+        try {
+            String filePath = "<YOUR_CREDENTIALS_FILE_PATH>";
+            String context = "abc";
+            SignedDataTokens signedToken = new SignedDataTokens.SignedDataTokensBuilder()
```
Comment (thefalc) [inline README.md:202]: Run through linter

File: README.md [hunk 16]
```diff
@@ -85,6 +88,165 @@ public class TokenGenerationUtil {
...
+                    .setTimeToLive(30.0) // in seconds
+                    .setDataTokens(new String[]{"dataToken1"}).build();
+
+            signedTokenValue =  signedToken.getSignedDataTokens();
+            System.out.println(signedTokenValue);
+        } catch (Exception e) {
+            e.printStackTrace();
+        }
+
+        // Generate BearerToken with context by specifying credentials.json as string
```
Comment (thefalc) [inline README.md:214]: Generate signed data tokens using a service account key string, context information, and a time to live.

File: README.md [hunk 17]
```diff
@@ -85,6 +88,165 @@ public class TokenGenerationUtil {
...
+        try {
+            String fileContents = "<YOUR_CREDENTIALS_FILE_CONTENTS_AS_STRING>";
+            String context = "abc";
+            SignedDataTokens signedToken = new SignedDataTokens.SignedDataTokensBuilder()
+                    .setCredentials(fileContents)
+                    .setContext(context)
+                    .setTimeToLive(30.0) // in seconds
```
Comment (thefalc) [inline README.md:221]: // Time to live set in seconds.

File: README.md [hunk 18]
```diff
@@ -85,6 +88,165 @@ public class TokenGenerationUtil {
...
+                    .setDataTokens(new String[]{"dataToken1"}).build();
+
+            signedTokenValue =  signedToken.getSignedDataTokens();
+            System.out.println(signedTokenValue);
+
+        } catch (SkyflowException e) {
+            e.printStackTrace();
+        }
+
```
Comment (thefalc) [inline README.md:230]: Remove these blank lines

File: README.md [hunk 19]
```diff
@@ -85,6 +88,165 @@ public class TokenGenerationUtil {
...
+
+    }
+}
+```
+Response:
+
+``` java
```
Comment (thefalc) [inline README.md:237]: Fix formatting. JSON should have 4 spaces for tabs

File: README.md [hunk 20]
```diff
@@ -85,6 +88,165 @@ public class TokenGenerationUtil {
...
+[
+   {
+      "token":“5530-4316-0674-5748“,
+      "signedToken":“signed_token_eyJhbGciOiJSUzI1NiJ9.eyJpc3MiOiJzLCpZjA“
+   }
+]
+```
+Note: 
+- Either credentials file path or credentials as string can be passed to `setCredentials` method of SignedDataTokensBuilder class.
```
Comment (thefalc) [inline README.md:246]: Update based on prior note.

File: README.md [hunk 21]
```diff
@@ -85,6 +88,165 @@ public class TokenGenerationUtil {
...
+- If both credentials file path and credentials string are passed to `setCredentials` method, then setter which is at the last will be taken precedence over the other.
+- `time-to-live` value is set in seconds.
```
Comment (thefalc) [inline README.md:248]: Time to live value expects time as seconds.

File: README.md [hunk 22]
```diff
@@ -85,6 +88,165 @@ public class TokenGenerationUtil {
     }
 }
 ```
+## Service Account Bearer Token with Context Generation
```
Comment (thefalc) [inline README.md:91]: I think some explanation of context is needed here or a link to the appropriate documentation.

**Review by thefalc — CHANGES_REQUESTED** [review]
Comments inline.

Let's also update the header.

# Skyflow Java
The Skyflow Java SDK is designed to help with integrating Skyflow into a Java backend. 

Update the ## Features list

- Authenticate with a Skyflow service account and generate a bearer token.
- Insert, retrieve and tokenize sensitive data.
- Invoke connections to call downstream third party APIs without directly handling sensitive data.

File: README.md (hunk 2 above)
Comment (manny-silva) [reply README.md:93]: To add onto this, what is `context`, and why would I want it?

File: README.md [hunk 23]
```diff
@@ -85,6 +88,165 @@ public class TokenGenerationUtil {
...
+            e.printStackTrace();
+        }
+
+    }
+}
+
+```
+
+Note: 
+- Either credentials file path or credentials as string can be passed to `setCredentials` method of BearerTokenBuilder class.
```
[the remark this answers is not in the document]
Comment (manny-silva) [reply README.md:143]: "You can pass either a service account key credentials file path or a service account key credentials as string to the `setCredentials` method of the BearerTokenBuilder class."

File: README.md (hunk 7 above)
Comment (manny-silva) [reply README.md:144]: "If you pass both a file path and string to the `setCredentials` method, the last method used takes precedence."

File: README.md [hunk 24]
```diff
@@ -85,6 +88,166 @@ public class TokenGenerationUtil {
...
+- If both credentials file path and credentials string are passed to `setCredentials` method, then setter which is at the last will be taken precedence over the other.
+- To generate multiple bearer tokens using a thread, please refer to this [Example](https://github.com/skyflowapi/skyflow-java/blob/master/samples/src/main/java/com/example/BearerTokenGenerationUsingThreadsExample.java)
```
Comment (manny-silva) [inline README.md:145]: "To generate multiple bearer tokens using a thread, see [this example](https://github.com/skyflowapi/skyflow-java/blob/master/samples/src/main/java/com/example/BearerTokenGenerationUsingThreadsExample.java)."

File: README.md [hunk 25]
```diff
@@ -85,6 +88,166 @@ public class TokenGenerationUtil {
...
+
+## Service Account Scoped Bearer Token Generation
+
+A service account having multiple roles can be used to generate bearer tokens with restricted access by providing the appropriate `roleID`. The token generated is valid for 60 minutes and perform opertaions with permissions associated with the role.
```
Comment (manny-silva) [inline README.md:149]: "A service account that has multiple roles can generate bearer tokens with access restricted to a specific role by providing the appropriate `roleID`."

File: README.md (hunk 8 above)
Comment (manny-silva) [reply README.md:149]: "Generated bearer tokens are valid for 60 minutes and can only perform operations with the permissions associated with the specified role."

File: README.md [hunk 26]
```diff
@@ -85,6 +88,166 @@ public class TokenGenerationUtil {
...
+            System.out.println(scopedToken);
+        } catch (Exception e) {
+            e.printStackTrace();
+        }
+    }
+}
+```
+Note: 
+- Either credentials file path or credentials as string can be passed to `setCredentials` method of BearerTokenBuilder class.
+- If both credentials file path and credentials string are passed to `setCredentials` method, then setter which is at the last will be taken precedence over the other.
```
Comment (manny-silva) [inline README.md:179]: Update based on comment in previous section.

File: README.md [hunk 27]
```diff
@@ -85,6 +88,166 @@ public class TokenGenerationUtil {
...
+[
+   {
+      "token":“5530-4316-0674-5748“,
+      "signedToken":“signed_token_eyJhbGciOiJSUzI1NiJ9.eyJpc3MiOiJzLCpZjA“
+   }
+]
+```
+Note: 
+- Either credentials file path or credentials as string can be passed to `setCredentials` method of SignedDataTokensBuilder class.
+- If both credentials file path and credentials string are passed to `setCredentials` method, then setter which is at the last will be taken precedence over the other.
```
Comment (manny-silva) [inline README.md:248]: Update based on previous comment.

**Review by manny-silva — CHANGES_REQUESTED** [review]

File: README.md (hunk 22 above)
Comment (Shaik-Luqmaan) [reply README.md:91]: Currently, we don't have any documentation in the doc site w.r.t CAA feature, will create a ticket for this in backlog and add the links once the documentation is updated.

File: README.md [hunk 28]
```diff
@@ -85,6 +88,165 @@ public class TokenGenerationUtil {
...
+            System.out.println(bearerToken);
+
+        } catch (SkyflowException e) {
+            e.printStackTrace();
+        }
+
+    }
+}
+
+```
```
[the remark this answers is not in the document]
Comment (Shaik-Luqmaan) [reply README.md:140]: This is the syntax to add Java sample code in .md file

File: README.md (hunk 1 above)
Comment (Shaik-Luqmaan) [reply README.md:105]: Done

File: README.md (hunk 11 above)
Comment (Shaik-Luqmaan) [reply README.md:164]: Done

File: README.md (hunk 15 above)
Comment (Shaik-Luqmaan) [reply README.md:203]: Done

File: README.md (hunk 19 above)
Comment (Shaik-Luqmaan) [reply README.md:238]: Done

**Pull request merged**